Add a range argument to vec.extract_if

the8472 opened the pull request:
tracking issue: #43244

This adds the range argument requested in https://github.com/rust-lang/rust/issues/43244#issuecomment-2486160659

the8472 committed:
Add a range argument to vec.extract_if

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11942424264/job/33289447588) [(plain)](https://github.com/rust-lang/rust/commit/34ed582fb094568adc4ad3b3b33751df48cf11df/checks/33289447588/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 27.4s done
#16 DONE 33.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error[E0422]: cannot find struct, variant or union type `Range` in this scope
  --> alloc/src/vec/extract_if.rs:49:13
   |
49 |         let Range { start, end } = slice::range(range, ..old_len);
   |
help: consider importing one of these structs
   |
1  + use crate::collections::btree_map::Range;
---

error[E0424]: expected value, found module `self`
  --> alloc/src/vec/extract_if.rs:55:26
   |
47 |     pub(super) fn new(vec: &mut Vec<T, A>, pred: F, range: core::ops::RangeFull) -> Self {
   |                   --- this function doesn't have a `self` parameter
...
55 |         ExtractIf { vec: self, idx: start, del: 0, end, old_len, pred: filter }
   |                          ^^^^ `self` value is a keyword only available in methods with a `self` parameter
   |
help: add a `self` receiver parameter to make the associated `fn` a method
   |
47 |     pub(super) fn new(&self, vec: &mut Vec<T, A>, pred: F, range: core::ops::RangeFull) -> Self {

error[E0425]: cannot find value `filter` in this scope
  --> alloc/src/vec/extract_if.rs:55:72
   |
   |
55 |         ExtractIf { vec: self, idx: start, del: 0, end, old_len, pred: filter }

error[E0425]: cannot find value `pred` in this scope
    --> alloc/src/vec/mod.rs:3690:30
     |
     |
3690 |         ExtractIf::new(self, pred, range)

error[E0308]: mismatched types
    --> alloc/src/vec/mod.rs:3690:36
     |
     |
3685 |     pub fn extract_if<F, R>(&mut self, range: R, filter: F) -> ExtractIf<'_, T, F, A>
...
...
3690 |         ExtractIf::new(self, pred, range)
     |         --------------             ^^^^^ expected `RangeFull`, found type parameter `R`
     |         arguments to this function are incorrect
     |
     = note:      expected struct `RangeFull`
             found type parameter `R`
             found type parameter `R`
note: associated function defined here
    --> alloc/src/vec/extract_if.rs:47:19
     |
47   |     pub(super) fn new(vec: &mut Vec<T, A>, pred: F, range: core::ops::RangeFull) -> Self {

Some errors have detailed explanations: E0308, E0422, E0424, E0425.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `alloc` (lib) due to 5 previous errors
```

</details>

tgross35 commented on library/alloc/src/vec/mod.rs:
````diff
@@ -3654,25 +3663,29 @@ impl<T, A: Allocator> Vec<T, A> {
+    /// ```
+    /// #![feature(extract_if)]
+    /// let mut items = vec![0, 0, 0, 0, 0, 0, 0, 1, 2, 1, 2, 1, 2];
+    /// let ones = items.extract_if(7.., |x| *x == 1).collect::<Vec<_>>();
````
Maybe change one of the first few `0`s to `1` to indicate they don't get extracted (out of range)?

the8472 replied:
Hrm, I wanted to show the "I have a prefix that I know doesn't contain those things, so I can skip them" use.

the8472 committed:
remove bounds from vec and linkedlist ExtractIf

since drain-on-drop behavior was removed those bounds
no longer serve a purpose

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11942670305/job/33290217933) [(plain)](https://github.com/rust-lang/rust/commit/007650475141123e64691ce95562ab07d4fcc4a2/checks/33290217933/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 28.1s done
#16 DONE 41.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling rustc_hir v0.0.0 (/checkout/compiler/rustc_hir)
   Compiling rustc_lint_defs v0.0.0 (/checkout/compiler/rustc_lint_defs)
   Compiling rustc_hir_pretty v0.0.0 (/checkout/compiler/rustc_hir_pretty)
   Compiling rustc_errors v0.0.0 (/checkout/compiler/rustc_errors)
error[E0277]: the trait bound `{closure@compiler/rustc_errors/src/lib.rs:1561:43: 1561:62}: RangeBounds<usize>` is not satisfied
     |
     |
1572 |                 diagnostic.children.extract_if(already_emitted_sub).for_each(|_| {});
     |                                     ---------- ^^^^^^^^^^^^^^^^^^^ the trait `RangeBounds<usize>` is not implemented for closure `{closure@compiler/rustc_errors/src/lib.rs:1561:43: 1561:62}`
     |                                     required by a bound introduced by this call
     |
     |
note: required by a bound in `Vec::<T, A>::extract_if`
     |
     |
3685 |     pub fn extract_if<F, R>(&mut self, range: R, filter: F) -> ExtractIf<'_, T, F, A>
...
...
3688 |         R: RangeBounds<usize>,
     |            ^^^^^^^^^^^^^^^^^^ required by this bound in `Vec::<T, A>::extract_if`
error[E0061]: this method takes 2 arguments but 1 argument was supplied
    --> compiler/rustc_errors/src/lib.rs:1572:37
     |
     |
1572 |                 diagnostic.children.extract_if(already_emitted_sub).for_each(|_| {});
     |                                     ^^^^^^^^^^--------------------- argument #2 is missing
note: method defined here
    --> /checkout/library/alloc/src/vec/mod.rs:3685:12
     |
     |
3685 |     pub fn extract_if<F, R>(&mut self, range: R, filter: F) -> ExtractIf<'_, T, F, A>
help: provide the argument
     |
     |
1572 |                 diagnostic.children.extract_if(already_emitted_sub, /* filter */).for_each(|_| {});

Some errors have detailed explanations: E0061, E0277.
For more information about an error, try `rustc --explain E0061`.
error: could not compile `rustc_errors` (lib) due to 2 previous errors
```

</details>

cuviper commented on library/alloc/src/vec/extract_if.rs:
```diff
@@ -24,12 +24,13 @@ pub struct ExtractIf<
-    F: FnMut(&mut T) -> bool,
+>
 {
     pub(super) vec: &'a mut Vec<T, A>,
```
Do we still need all these fields to be `pub(super)` now that there's a local `fn new`?

the8472 replied:
removed the visibility.

the8472 committed:
update uses of extract_if in the compiler

the8472 committed:
remove obsolete comment and pub(super) visibility

the8472 commented on library/alloc/src/vec/extract_if.rs:
```diff
@@ -82,14 +92,12 @@ where
-    F: FnMut(&mut T) -> bool,
 {
     fn drop(&mut self) {
         unsafe {
```
removed the comment.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11943156162/job/33291672270) [(plain)](https://github.com/rust-lang/rust/commit/5d0e0a7c4c3439fc8c71c577060bc7b8d222f42b/checks/33291672270/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
     = note: `-D rustdoc::unescaped-backticks` implied by `-D warnings`
     = help: to override `-D warnings` add `#[allow(rustdoc::unescaped_backticks)]`
help: if you meant to use a literal backtick, escape it
     |
3658 |     /// If \`range`` is out of bounds.

error: could not document `alloc`
warning: build failed, waiting for other jobs to finish...
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Wed Nov 20 23:20:52 UTC 2024
  network time: Wed, 20 Nov 2024 23:20:52 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

cuviper commented on library/alloc/tests/vec.rs:
```diff
@@ -1492,6 +1492,31 @@ fn extract_if_true() {
+
+#[test]
+#[should_panic]
+fn extraxt_if_out_of_bounds() {
```
```suggestion
fn extract_if_out_of_bounds() {
```

cuviper reviewed (approved):
LGTM.

The FCP finished -- do you want to go ahead and add the stabilization too?

tgross35 commented:
r? @cuviper (sorry, I didn't realize this was assigned to me)

> The FCP finished -- do you want to go ahead and add the stabilization too?

Since this is a pretty significant API and implementation change, maybe it would be good to hold off for a couple of weeks? In case anyone using this API unstably has feedback after the change (I doubt it, but doesn't hurt).

the8472 commented:
Changing and stabilizing in the same PR doesn't seem ideal to me either, especially when the change happened parallel to the FCP. I think it should bake a few days at least.

cuviper commented:
@bors r+

bors commented:
:pushpin: Commit 414acdad1cc996fe90c4e866f20df4d6022fe399 has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper 414acdad1cc996fe90c4e866f20df4d6022fe399 -->
<!-- homu: {"type":"Approved","sha":"414acdad1cc996fe90c4e866f20df4d6022fe399","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #133522) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

cuviper commented:
I think you forgot the "extraxt" typo fix when you rebased.

the8472 commented:
Indeed, fixed.

@bors r=cuviper

bors commented:
:pushpin: Commit 44790c490902d3eb5871a13b80ac4d6c19909278 has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper 44790c490902d3eb5871a13b80ac4d6c19909278 -->
<!-- homu: {"type":"Approved","sha":"44790c490902d3eb5871a13b80ac4d6c19909278","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

nk9 commented on library/alloc/src/vec/mod.rs:
````diff
@@ -3630,10 +3633,12 @@ impl<T, A: Allocator> Vec<T, A> {
     /// ```
+    /// # use std::cmp::min;
     /// # let some_predicate = |x: &mut i32| { *x == 2 || *x == 3 || *x == 6 };
     /// # let mut vec = vec![1, 2, 3, 4, 5, 6];
````
Apologies for the late comment, but I wanted to point out that this sample code doesn't behave the same way as the real `extract_if()`. Given the same predicate and range, if your vector were `[1, 2, 3, 3, 3, 3, 3, 3, 4, 5, 6]`, then all of the 3s would be removed. `i` is only incremented when an element is dropped, but `range.end` is unchanged, so the items shift down. I got very confused when reading the docs and trying to square this sample code with the explanation of how the function works.

Fortunately, the real `extract_if()` does not have this problem. Proposed change to the sample code:

```rust
    let some_predicate = |x: &mut i32| { *x == 2 || *x == 3 || *x == 5 };
    let mut vec = vec![0, 1, 2, 3, 4, 5, 6];
    let range = 1..5;
    let mut i = range.start;
    let mut end = range.end;
    while i < min(vec.len(), end) {
        if some_predicate(&mut vec[i]) {
            let _val = vec.remove(i);
            // your code here
            end -= 1;
        } else {
            i += 1;
        }
    }


    assert_eq!(vec, vec![0, 1, 4, 5, 6]);
```

the8472 replied:
If you want to propose a documentation improvement you can just open a PR.

nk9 reviewed (commented):
I hope this is the right place to raise this question about the sample code in this PR.
